### PR TITLE
XAML QuickInfo update

### PR DESCRIPTION
**Pull request by mgoertz-msft**
Updated XAML QuickInfo to show more info like C# by using ISymbolDisplayService and adding more documentation parts.

Before:
QuickInfo-Before

After:
QuickInfo-After

### CODE REVIEW

**Comment by sharwell** [comment]
Can you add before/after screenshots?

**Comment by mgoertz-msft** [comment]
@sharwell Done. See updated description.

**Comment by sharwell** [comment]
Nice! Does the hyperlinking work as well?

**Comment by mgoertz-msft** [comment]
No, it does not. This is for the LSP Hover handler, which I don't think supports that.

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs [hunk 1]
```diff
@@ -53,16 +52,51 @@ public static async Task<IEnumerable<TaggedText>> GetDescriptionAsync(this ISymb
                 return Enumerable.Empty<TaggedText>();
             }
 
-            var textContentBuilder = new List<TaggedText>();
-            textContentBuilder.AddRange(symbol.ToDisplayParts(s_descriptionStyle).ToTaggedText());
+            var description = await symbolDisplayService.ToDescriptionPartsAsync(codeProject.Solution.Workspace, semanticModel, 0, ImmutableArray.Create(symbol), SymbolDescriptionGroups.MainDescription, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs:55]: named parameter for 0.

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs [hunk 2]
```diff
@@ -53,16 +52,51 @@ public static async Task<IEnumerable<TaggedText>> GetDescriptionAsync(this ISymb
...
+
+            var builder = new List<TaggedText>();
+            builder.AddRange(description.ToTaggedText());
 
-            var documentation = symbol.GetDocumentationParts(semanticModel, offset, formatter, cancellationToken);
+            var documentation = symbol.GetDocumentationParts(semanticModel, 0, formatter, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs:60]: named parameter for 0.

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs [hunk 3]
```diff
@@ -53,16 +52,51 @@ public static async Task<IEnumerable<TaggedText>> GetDescriptionAsync(this ISymb
...
             if (documentation.Any())
             {
-                textContentBuilder.AddLineBreak();
-                textContentBuilder.AddRange(documentation);
+                builder.AddLineBreak();
+                builder.AddRange(documentation);
+            }
+
+            var remarksDocumentation = symbol.GetRemarksDocumentationParts(semanticModel, 0, formatter, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs:67]: i'm surprpised you need to write this yourself.  if you use ToDescriptionGroupsAsync, you should have a documentation group already setup for you that you can use.  does that not work?

**Comment by CyrusNajmabadi** [comment]
Overall, this looks great.  Just a general question about if we can make it even simpler by leveraging more of what ISYmbolDisplayService gives you :)

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs (hunk 3 above)
Comment (mgoertz-msft) [reply src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs:67]: Nope. I would have expected ToDescriptionPartsAsync with SymbolDescriptionGroups.All to give me everything but it does not, so I limited it to MainDescription and called the rest myself.

**Comment by mgoertz-msft** [comment]
Thanks @CyrusNajmabadi. I agree ISymbolDisplayService should have been able to do all this for me. Right now, it does not make a difference if I pass in All groups or just MainDescription to ToDescriptionPartsAsync, both just give me the full symbol name.

**Comment by mgoertz-msft** [comment]
Thanks @CyrusNajmabadi. I agree ISymbolDisplayService should have been able to do all this for me. Right now, it does not make a difference if I pass in All groups or just MainDescription to , both just give me the full symbol name.

**Comment by mgoertz-msft** [comment]
@sharwell, @CyrusNajmabadi, anything else you need before signing off?

**Comment by CyrusNajmabadi** [comment]
Sorry. I need to look into this. I have to understand why we can't currently share more.

**Comment by mgoertz-msft** [comment]
@CyrusNajmabadi How about we go with this for now and I open an issue?

**Comment by CyrusNajmabadi** [comment]
looking now.

**Comment by CyrusNajmabadi** [comment]
I've opened https://github.com/dotnet/roslyn/pull/51961 to help out here.

**Comment by mgoertz-msft** [comment]
Thanks @CyrusNajmabadi! I'll wait for it. Looks like #51961 has some (unrelated?) test failures though.

**Comment by CyrusNajmabadi** [comment]
@mgoertz-msft my pr went in.  So you should be able to use that and just combine the groups without having to rebuild the internal structure.  Thanks!

**Comment by mgoertz-msft** [comment]
@CyrusNajmabadi Shouldn't I be able to call ToDescriptionPartsAsync with DescriptionGroups.All to get everything? That still doesn't work because the builder only puts the MainDescription into the _groupsMap and the rest into the _documentationMap.

**Comment by CyrusNajmabadi** [comment]
Call ToDescriptionGroupsAsync instead.  That gives you the merged result.  You shouldn't need SymbolDisplayParts (i hope) for what you're doing.  TaggedText is the better, and richer, medium for doing things like QuickInfo.

**Comment by mgoertz-msft** [comment]
I would have just called ToTaggedText() on the result. With ToDescriptionGroupsAsync I have to iterate over all the groups and merge the results myself. Isn't it still a bug that ToDescriptionPartsAsync takes a group flag as an argument but only ever supports the MainDescription?

**Comment by CyrusNajmabadi** [comment]
>I would have just called ToTaggedText() on the result. With ToDescriptionGroupsAsync I have to iterate over all the groups and merge the results myself. 

We can make an extension that does that for you if htat would help :)

> Isn't it still a bug that ToDescriptionPartsAsync takes a group flag as an argument but only ever supports the MainDescription?

In this case no.  Because we can't actually make SymbolDisplayParts for these other sections.  Specifically, these other sections use features of TaggedText that are not representable in the SYmbolDisplayPart world.  SymbolDisplayParts are a subset of TaggedText.

This is very much because this is an api using both compiler and IDE primitives, and it has evolved a lot over the years.  it likely could be cleaned up, but it was internal and we never had the need.

**Comment by CyrusNajmabadi** [comment]
i think you'll want to extract the code in CommonSemanticQuickInfoProvider.CreateContentAsync where we grab all this data and merge it together and then use that common code for your new provider.

**Comment by CyrusNajmabadi** [comment]
actually, let me try to provide the helper.  i'll make a pr for thsi.

**Comment by mgoertz-msft** [comment]
@CyrusNajmabadi New QuickInfoUtilities helper works great, thanks, but I had to insert line breaks between the sections.

**Comment by mgoertz-msft** [comment]
@CyrusNajmabadi New QuickInfoUtilities helper works great, thanks, but I had to insert line breaks between the sections.

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs [hunk 4]
```diff
@@ -53,16 +54,17 @@ public static async Task<IEnumerable<TaggedText>> GetDescriptionAsync(this ISymb
                 return Enumerable.Empty<TaggedText>();
             }
 
-            var textContentBuilder = new List<TaggedText>();
-            textContentBuilder.AddRange(symbol.ToDisplayParts(s_descriptionStyle).ToTaggedText());
-
-            var documentation = symbol.GetDocumentationParts(semanticModel, offset, formatter, cancellationToken);
-            if (documentation.Any())
+            var quickInfo = await QuickInfoUtilities.CreateQuickInfoItemAsync(codeProject.Solution.Workspace, semanticModel, TextSpan.FromBounds(0, 0), ImmutableArray.Create(symbol), cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Extensions/SymbolExtensions.cs:57]:
```suggestion
            var quickInfo = await QuickInfoUtilities.CreateQuickInfoItemAsync(codeProject.Solution.Workspace, semanticModel, span: default, ImmutableArray.Create(symbol), cancellationToken).ConfigureAwait(false);
```

(up to you on what you prefer here).

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**